### DISCUSSION TITLE
Added: LDM Notes for May 3rd, 2021

### DISCUSSION THREAD

**Discussion by 333fred** [dotnet/csharplang#4720] [LDM Notes]
https://github.com/dotnet/csharplang/blob/main/meetings/2021/LDM-2021-05-03.md

## Agenda

1. Improved interpolated strings
2. Open questions in record structs

**Comment by alrz** [comment]
>For better codegen, we will go with a single method of creation, a constructor, not a Create method.

This part is a little unclear. How using a ctor vs method improves codegen?


>We will treat interpolated string builder usage as any other usage of a ref struct in async methods (today, that means disallowed).

Does that mean we'll emit string.Format in async code or just no awaits in holes? Either way that's unfortunate. I was looking forward to the resolution on this because enumerables list patterns will need some kind of buffer type, only that we won't be able to use a uniform ref API in both async/sync code. (relevant discussion: https://github.com/dotnet/csharplang/discussions/4687)

**Reply by 333fred** [reply]
> How using a ctor vs method improves codegen?

Various mechanisms around optimizations the JIT can do. As this is a feature focused on perf, that is something we want to take advantage of.

> Does that mean we'll emit string.Format in async code or just no awaits in holes?

The spec is and will be intentionally vague for interpolated string literals used directly as strings, to allow for flexible codegen in the compiler for this one specific case; we know _exactly_ the lifetimes involved with that codegen, so we can optimize without having to make general special cases, determining safety rules, and etc. I believe enumerable list patterns would fall in a similar area, as we know there can't be any awaits in the pattern, the struct isn't exposed to user code, and we guarantee the exact lifetime of the struct.
